### ISSUE TITLE
[Question] problem with delta parameter of savgol_filter() function of scipy.signal?

### ISSUE THREAD

**Issue by ThiagoDiasV** [scipy/scipy#9910]
Hello. I'm not too expert on data analysis but I'm having some problem with the savgol_filter() function from scipy.signal. 

I'm trying to use the Saviztky-Golay filter with differents delta's (delta=1.0, delta=2.0, delta=3.0 for instance) but I've noticed that this parameter is dividing the values obtained from Saviztky-Golay filter by the delta value, and this isn't the expected behavior of delta parameter. I've compared the data obtained using savgol_filter() function with data obtained from a statistics software and they don't fit. So, I tried to make some tests to understand what is wrong with the filter.

The documentation says: 

```
delta : float, optional

The spacing of the samples to which the filter will be applied. This is only used if deriv > 0. Default is 1.0.
```

And the derivative definition is: 

![image](https://user-images.githubusercontent.com/42140919/53960679-a60b3c80-40c5-11e9-9529-e288e0cefc13.png)

So, the expected behavior of delta parameter is make the subtraction of two points. If my series is like this: 
```
x = [212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223]
y = [0.450, 0.460, 0.470, 0.475, 0.483, 0.490, 0.501, 0.493, 0.485, 0.460, 0.450, 0.440]
```
The values of Savitzky-Golay filter applied with window_length = 11 and polyorder = 2 is:

```
>>> y_savgol = savgol_filter(y, 11, 2)
>>> y_savgol = [0.450179, 0.459023, 0.468946, 0.47827, 0.487163, 0.491744, 0.491779, 0.487417, 0.479212 0.467916, 0.45486, 0.441769]
```
If my delta = 2, for the values in index 4 (x=216, y_savgol=0.487163), for instance, we have: 

```
derivative_delta_2 = ((0.491779-0.487163)/(218-216) + (0.487163-0.468946)/(216-214))/2
derivative_delta_2 = 0.005708
```
But the result of the savgol_filter(y, window_length=11, polyorder=2, deriv=1, delta=2.0) in this exactly point is:
```
savgol_filter_derivative_delta_2 = 0.001891
```
Clearly they're different. 

So I tried to understand what's wrong. And I noticed that delta parameter makes this:
```
>>> delta_1 = savgol_filter(data, 11, 2, 1, 1.0) 
>>> delta_1[some_point] = x

>>> delta_2 = savgol_filter(data, 11, 2, 1, 2.0)
>>> delta_2[some_point] = x/2

>>> delta_3 = savgol_filter(data, 11, 2, 1, 3.0)
>>> delta_3[some_point] = x/3 
```
If I'm not doing anything wrong, I believe that this is a bug that can be fixed. 

### Scipy/Numpy/Python version information:

```
1.1.0 1.14.3 sys.version_info(major=3, minor=6, micro=5, releaselevel='final', serial=0)
```

**Comment by larsoner** [comment]
@WarrenWeckesser this looks like your code (from 6 years ago!), any ideas?

**Comment by WarrenWeckesser** [comment]
For your example of estimating the first derivative using a finite difference, you wrote

    derivative_delta_2 = ((0.491779-0.487163)/(218-216) + (0.487163-0.468946)/(216-214))/2

You should not use the values 0.491779, 0.487163, etc. for this calculation.  You must use the original `y` values.  To estimate the derivative at index 4, you could use something like this:

    delta = 2.0
    derivative_delta_2 = ((y[5] - y[4])/delta + (y[4] - y[3])/delta)/2

I get 0.00375 for that value:

    In [82]: y = [0.450, 0.460, 0.470, 0.475, 0.483, 0.490, 0.501, 0.493, 0.485, 0.460, 0.450, 0.440]

    In [83]: delta = 2.0

    In [84]: derivative_delta_2 = ((y[5] - y[4])/delta + (y[4] - y[3])/delta)/2

    In [85]: derivative_delta_2
    Out[85]: 0.0037500000000000033

This is almost exactly the value returned by `savgol_filter` if the window length is 3:

    In [87]: savgol_filter(y, window_length=3, polyorder=2, deriv=1, delta=2)[4]
    Out[87]: 0.0037499999999999925

Increasing the window length uses more data points, so the estimate is not expected to agree with a simple finite difference.

    In [88]: savgol_filter(y, window_length=5, polyorder=2, deriv=1, delta=2)[4]
    Out[88]: 0.0038500000000000556

    In [89]: savgol_filter(y, window_length=11, polyorder=2, deriv=1, delta=2)[4]
    Out[89]: 0.0021970862470862404

Note that `savgol_filter` knows nothing about the `x` values of your data.  The `delta` parameter specifies the spacing of x-coordinates of the given `y` values.  Based on the incorrect formula you gave for `derivative_delta_2`, I think you might have mistaken `delta` for an increment of the *index* of `y`.  That is not the meaning of `delta`.

Here's a example that might help clarify how `delta` is used.

```
import numpy as np
from scipy.signal import savgol_filter
import matplotlib.pyplot as plt


T = 10.0
fs = 500.0  # Hz
nsamples = int(T*fs)
t = np.arange(nsamples)/fs

x = np.sin(2*t)

# dxdt is the exact derivative.
dxdt = 2*np.cos(2*t)

# Estimate the derivative using the Savitzky-Golay filter.
# Note that delta is set to 1/fs, which is the sampling period.
y1 = savgol_filter(x, 11, 2, deriv=1, delta=1/fs)

plt.plot(t, dxdt, 'k--', alpha=0.2, linewidth=3, label='exact')
plt.plot(t, y1, 'b', alpha=0.75, linewidth=1, label='Savitzky-Golay')

plt.xlabel('t')
plt.legend(shadow=True, loc='lower right')
plt.show()
```
The plot shows that the estimate of the derivative works as expected:
![image](https://user-images.githubusercontent.com/321463/53985375-74a46800-40e9-11e9-8f9f-0488dda27721.png)

**Comment by WarrenWeckesser** [comment]
@ThiagoDiasV, I'm closing the issue, but don't hesitate to ask further questions if my comments above are not clear.

**Comment by ThiagoDiasV** [comment]
@WarrenWeckesser Thank you for your clear explanation. 

Now I see that I made a mistake. I was trying to apply savgol_filter() method on a derivative spectroscopy (http://www.whoi.edu/cms/files/derivative_spectroscopy_59633940_175744.pdf). 

The way that I explained was correct but for my method. I needed to apply the first derivative in my results after the use of savgol_filter(). Because of this, in my explanation, I tried to use the first derivative on the results of Saviztky-Golay filter. This approach is common on derivative spectroscopy. 

When I saw that one parameter of the method was delta parameter, I got mixed up and thought that what I was trying to do was possible directly using savgol_filter(). But when you said that the method doesn't know nothing about my x axis the things got clear to me (and I didn't have realized that, shame on me! haha). 

Now it's pretty clear to me that the delta of derivative spectroscopy is different from delta of savgol_filter() method. Their functions are different. 

Sorry about my mess. And thanks again for your explanation and your time.
